testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    cluster::Cluster,
+    health::{HealthCheck, HealthCheckContext},
+    prometheus::Prometheus,
+    util::unix_timestamp_now,
+};
+use anyhow::{format_err, Result};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::{collections::HashMap, time::Duration};
+
+static VAL_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"val-(\d+)").unwrap());
+static FULLNODE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"fn-(\d+)").unwrap());
+
+pub struct FullNodeHealthCheck {
+    cluster: Cluster,
+    prometheus: Prometheus,
+}
+
+impl FullNodeHealthCheck {
+    pub fn new(cluster: Cluster, prometheus: Prometheus) -> FullNodeHealthCheck {
+        Self {
+            cluster,
+            prometheus,
+        }
+    }
+
+    pub fn state_sync_committed_version(
+        &self,
+        start: Duration,
+        end: Duration,
+    ) -> Result<HashMap<String, f64>> {
+        libra_retrier::retry(libra_retrier::fixed_retry_strategy(1_000, 5), || {
+            self.prometheus
+                .query_range_max(
+                    "libra_state_sync_committed_version{{}}".to_string(),
+                    &start,
+                    &end,
+                    10, /* step */
+                )
+                .map_err(|e| format_err!("No sync committed version data: {}", e))
+        })
+    }
+}
+
+pub fn convert_index(input: HashMap<String, f64>) -> HashMap<String, Vec<i64>> {
```

Comment:
Returning negative index for validator and then sorting array seem to be very fragile and error prone

I think you can avoid convert_index function all together.

When you get `Map<String, f64>` from prometheus, you can do something like this:
```
for full_node in cluster.fullnode_instances() {
let full_node_data_point = map.get(full_node.peer_name());
let validator_data_point = map.get("val-" + full_node.validator_index()); // well use format! here, but you got the idea
// compare full_node_data_point and validator_data_point
}
```

Doing this way have few benefits over your approach
a) It is simpler
b) It allows to catch problems like if there is no data point for some full node or validator (this should fail health check, as this is not normal situation)